tracer/src/Datadog.Trace/AppSec/Waf/ReturnTypes.Managed/InitResult.cs

```diff
@@ -65,14 +64,37 @@ private InitResult(ushort failedToLoadRules, ushort loadedRules, string ruleFile
 
         internal bool Reported { get; set; }
 
-        internal static InitResult FromUnusableRuleFile() => new(0, 0, string.Empty, new Dictionary<string, string[]>(), unusableRuleFile: true);
+        internal static InitResult FromUnusableRuleFile() => new(0, 0, string.Empty, new Dictionary<string, object>(), unusableRuleFile: true);
 
-        internal static InitResult From(DdwafRuleSetInfo ddwaRuleSetInfo, IntPtr? wafHandle, WafLibraryInvoker wafLibraryInvoker)
+        internal static InitResult From(IntPtr diagnostics, IntPtr? wafHandle, WafLibraryInvoker? wafLibraryInvoker)
         {
-            var ddwafObjectStruct = ddwaRuleSetInfo.Errors;
-            var errors = ddwafObjectStruct.Decode();
-            var ruleFileVersion = Marshal.PtrToStringAnsi(ddwaRuleSetInfo.Version);
-            return new(ddwaRuleSetInfo.Failed, ddwaRuleSetInfo.Loaded, ruleFileVersion!, errors, wafHandle: wafHandle, wafLibraryInvoker: wafLibraryInvoker);
+            ushort failedCount = 0;
+            ushort loadedCount = 0;
+            string rulesetVersion = string.Empty;
+            Dictionary<string, object>? errors = null;
+            try
+            {
+                if (diagnostics != IntPtr.Zero)
+                {
+                    var diagnosticsData = (Dictionary<string, object>)Encoder.Decode(new Obj(diagnostics));
+                    if (diagnosticsData.Count > 0)
+                    {
+                        var rules = (Dictionary<string, object>)diagnosticsData["rules"];
```

Comment:
this crashes silently in the RCM updates. If you look at this PR artifacts you'll find:

```
2023-07-05 17:35:42.702 +00:00 [WRN] AppSec could not read Waf diagnostics. Disabling AppSec : System.Collections.Generic.KeyNotFoundException: The given key 'rules' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Datadog.Trace.AppSec.Waf.ReturnTypes.Managed.InitResult.From(IntPtr diagnostics, Nullable`1 wafHandle, WafLibraryInvoker wafLibraryInvoker) in /_/tracer/src/Datadog.Trace/AppSec/Waf/ReturnTypes.Managed/InitResult.cs:line 82  { MachineName: ".", Process: "[5584 dotnet]", AppDomain: "[1 Samples.Security.AspNetCore5]", AssemblyLoadContext: "\"Default\" System.Runtime.Loader.DefaultAssemblyLoadContext #2", TracerVersion: "2.34.0.0" }
```

this is because we cant expect this keys to be always here, they wont be for RCM updates but we need to read other ones like rule-overrides. 
And general I think we should now decorrelate `InitResult` from `UpdateResult` and we should stop in `UpdateResult` doing ` var result = InitResult.From(diagnostics, null, null);` because we are not expecting the same keys anymore. The integration test passes here because it's logging a warning in the catch(exception).

It doesnt pass [here ](https://github.com/DataDog/dd-trace-dotnet/pull/4372) because we are explicitely looging errors if we dont find the keys in `InitResult`. 

I think InitResult should continue to try look for a `rules` key and log an error if not (like in the minor fixes sub PR #4372) and UpdateResult should read the keys in another way, to account for updates and to not expect rules necessarily.